### PR TITLE
[None][refactor] Modularize sparse Top-K selection

### PR DESCRIPTION
**Pull request by lfr-0531**
## Dev Engineer Review

- Added the reusable `TopK` module for Torch, CUDA Radix, CuTe DSL Radix, CUDA GVR, and CuTe DSL GVR.
- Centralized GVR state and reusable Radix/GVR workspace management.
- Updated DSA Indexer and TriAttention to use shared Top-K dispatch.
- Preserved the CUDA Radix fallback for compressed MTP decode.
- Removed the obsolete `use_custom_topk` API and heuristic warmup export.
- Updated DSA metadata, warmup handling, and prior-index state preservation.
- Review focus: backend fallback conditions, buffer ownership, row-bound handling, and compatibility with callers of the removed `use_custom_topk` argument.
- CI later completed successfully for commit `6edec12` in pipeline `#54896`.

## QA Engineer Review

- Added `TopK` unit tests for Torch prefill/decode behavior, CuTe DSL Radix fallback, GVR state handling, CUDA Radix dispatch, CUDA GVR workspace management, unsupported configurations, and error handling.
- Updated DSA Indexer tests for Top-K selection, warmup dispatch, shared state, prefill/decode paths, cached requests, MTP paths, draft replay, and row ordering.
- Test-list coverage in `tests/integration/test_lists/` is not confirmed.
- Verdict: needs follow-up.

## Description

Sparse attention currently selects Top-K through backend-specific branches in the DSA Indexer, while GVR prior indices and scratch buffers live in shared attention metadata. TriAttention also calls the CuTe DSL Top-K operation directly. This makes implementation selection and state ownership fragmented across callers.

This PR introduces a reusable internal `TopK` module with Torch, CUDA Radix, CuTe DSL Radix, CUDA GVR, and CuTe DSL GVR implementations. The module provides one prefill/decode entry point, owns GVR runtime state, and preserves the CUDA Radix fallback for compressed MTP decode. DSA Indexer and TriAttention now instantiate and call this module instead of implementing their own dispatch. The existing model-engine warmup flow from `main` remains unchanged.

This builds on the sparse-attention backend refactor merged in #12733. The DSA, TriAttention, and test changes are included together because they are the production consumers of the shared module.

## Test Coverage

- `tests/unittest/_torch/modules/test_top_k.py`: 8 passed
- DSA metadata warmup dispatch and Indexer/TopK ownership tests: 6 passed
- `test_triattention_selection_compaction.py::test_union_eager_cuda_resolves_heavy_ties_and_ragged_lengths[4-64]`: 1 passed
- Changed-file formatting, lint, Python syntax, DCO, and test-list AST validation passed

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/modules/top_k.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+            prior_indices[: self._gvr_prior_indices.shape[0]].copy_(self._gvr_prior_indices)
+            self._gvr_prior_indices = prior_indices
+
+        if self.decode_implementation == TopKImplementation.CUDA_GVR:
+            if (
+                self._cuda_gvr_scratch.dtype != scores.dtype
+                or self._cuda_gvr_scratch.shape[0] < row_capacity
+            ):
+                self._cuda_gvr_scratch = scores.new_empty((row_capacity, self.top_k))
+            return
```
Comment (lori-ren) [inline tensorrt_llm/_torch/modules/top_k.py]: The resizing and re-assignment of local buffer within a cuda graph path will break the graph silently with significant performance downgrade. Please check.

Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py]: Fixed. Thanks Lori! In `ee35f683f9`, I removed that logic and the `_ensure_gvr_buffers()` method entirely. The persistent GVR prior is now allocated once by DSA metadata with shape `(num_local_layers, max_num_sequences, top_k)`, which gives each layer a fixed-address slice. GVR and radix scratch use the existing reusable CUDA memory-buffer arena; buffers requested during capture are marked reserved so they are not reclaimed or reused by another graph.

File: tensorrt_llm/_torch/modules/top_k.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    def _get_radix_workspace(
+        self, scores: torch.Tensor
+    ) -> tuple[torch.Tensor | None, torch.Tensor | None]:
+        if scores.dtype != torch.float32:
+            return None, None
+
+        shape = (scores.shape[0], _MAX_RADIX_BLOCKS_PER_ROW, self.top_k)
+        capture_graph = scores.is_cuda and torch.cuda.is_current_stream_capturing()
+        radix_indices = self._memory_buffers.get_buffer(
```
Comment (mingyangHao) [inline tensorrt_llm/_torch/modules/top_k.py]: `TopK._memory_buffers` is process-global, but buffers are keyed only by name and allocated on the current `"cuda"` device. Calling Top-K on multiple GPUs can therefore reuse a workspace from the wrong device and fail; please key and allocate these workspaces using `scores.device`.

File: tensorrt_llm/_torch/modules/top_k.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                )
+            row_order = None
+            if num_requests * next_n >= 2 * self._num_sms:
+                row_order = self._memory_buffers.get_buffer(
+                    (num_requests,),
+                    dtype=torch.int32,
+                    buffer_name="top_k_cute_dsl_gvr_row_order",
+                    reserve_buffer=capture_graph,
+                )
+                row_order.copy_(torch.argsort(sequence_lengths, descending=True).to(torch.int32))
```
Comment (mingyangHao) [inline tensorrt_llm/_torch/modules/top_k.py]: This sorts the same `sequence_lengths` once per full-indexer layer, adding repeated work in eager execution and CUDA graphs. Please compute the row order once per forward step, as the previous metadata path did, and reuse it across layers.

File: tests/unittest/_torch/modules/test_top_k.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (lori-ren) [inline tests/unittest/_torch/modules/test_top_k.py:1]: It seems that this test file is not included in any test list. Please add.

**Review by lori-ren — APPROVED** [review]

**Review by limin2021 — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]

File: tests/unittest/_torch/modules/test_top_k.py (hunk 4 above)
Comment (lfr-0531) [reply tests/unittest/_torch/modules/test_top_k.py:1]: This file is already covered by the `unittest/_torch/modules` directory entry in `tests/integration/test_lists/test-db/l0_cpu.yml`, so pytest recursively discovers every test in `test_top_k.py`.

File: tensorrt_llm/_torch/modules/top_k.py (hunk 2 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py]: Fixed. Thanks, Mingyang! TopK now includes `scores.device` in every reusable-workspace key and enters the corresponding CUDA device context before checking capture state and requesting storage from the process-global arena.

File: tensorrt_llm/_torch/modules/top_k.py (hunk 3 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py]: Fixed by restoring row-order preparation to DSA metadata. The descending argsort is now computed once in the per-forward metadata update, written into a metadata buffer, and passed to every layer's TopK call.

**Review by mingyangHao — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py [hunk 5]
```diff
@@ -699,11 +637,22 @@ def __init__(
...
+            decode_top_k_implementation = (
+                TopKImplementation.CUTE_DSL_GVR
+                if self._enable_heuristic_topk
+                else TopKImplementation.CUTE_DSL_RADIX
+            )
+        elif self._enable_heuristic_topk:
+            decode_top_k_implementation = TopKImplementation.CUDA_GVR
+        else:
+            decode_top_k_implementation = TopKImplementation.CUDA_RADIX
+        self.top_k = TopK(
```
Comment (longcheng-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py:650]: `warmup_heuristic_topk_decode` existed to keep the C++ dispatcher's lazy `cudaDeviceGetAttribute` calls (SM count, L2 size) out of CUDA graph capture. The pre-capture warmup forwards should still populate those caches, but after this removal that is an implicit guarantee. Can you confirm every capture path runs an eager `indexer_topk_decode` first?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py [hunk 6]
```diff
@@ -1577,55 +1506,29 @@ def sparse_attn_indexer(
...
-            local_layer = metadata.kv_cache_manager.layer_offsets[self.layer_idx]
-            ctx_seq_lens = metadata.seq_lens[:num_contexts]
-            # Per-sequence last context-token offset (exclusive cumsum minus 1).
-            last_ctx_idx = (torch.cumsum(ctx_seq_lens, dim=0) - 1).to(dtype=torch.long)
-            metadata.heuristic_prev_topk[
-                local_layer, num_generations : num_generations + num_contexts
-            ].copy_(topk_indices_buffer[last_ctx_idx, :])
+        # Update GVR state after chunk all-gathers or the dense-index copy.
+        if has_prefill:
+            self.top_k.update_gvr_prior_from_prefill(
```
Comment (longcheng-nv) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py:1526]: Behavior change vs main: the prior update now also runs on the skip_indexer paths (main gated the seeding on `not metadata.skip_indexer_for_ctx_reqs`, and the decode write-back at L1734 lived inside the non-skip branch), so dense indices get written into `gvr_prior_indices`. Harmless today: GVR admission needs `numColumns >= 12288`, far above the skip threshold (`kv_len <= K`), so the dense hint is never consumed. But that relies on the admission floor staying well above K, and `TRTLLM_HEURISTIC_NMIN` can lower it to 1024. If the simplification is intentional, please note that assumption in a comment.

File: tensorrt_llm/_torch/modules/top_k.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        else:
+            torch.ops.trtllm.cute_dsl_gvr_topk_decode(
+                scores,
+                gvr_prior_indices,
+                sequence_lengths,
+                output_indices,
+                self.top_k,
+                next_n=next_n,
+                compress_ratio=self.compress_ratio,
+                max_seq_len=scores.shape[1],
```
Comment (longcheng-nv) [inline tensorrt_llm/_torch/modules/top_k.py]: main passes `indexer_max_seq_len` here; `scores.shape[1]` only matches if the decode logits buffer is always allocated at full width. `max_seq_len` is a DSL compile key, so a narrower per-step view would trigger extra JITs or miss the warmed key. Is `shape[1]` guaranteed constant across steps?

File: tensorrt_llm/_torch/modules/top_k.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            shape,
+            dtype=dtype,
+            buffer_name=device_buffer_name,
+            reserve_buffer=False,
+        )
+
+    def _get_radix_workspace(
+        self, scores: torch.Tensor
+    ) -> tuple[torch.Tensor | None, torch.Tensor | None]:
+        if scores.dtype != torch.float32:
```
Comment (longcheng-nv) [inline tensorrt_llm/_torch/modules/top_k.py:230]: Returning `(None, None)` for non-fp32 scores means the C++ op aborts if a bf16 caller ever needs `blocks_per_row > 1` (main allocated the aux buffers unconditionally). Dead path today since indexer logits are fp32, but an assert or a comment would keep this from failing at a distance later.

**Review by longcheng-nv — COMMENTED** [review]
Reviewed with a focus on GVR semantics. The workspace move to `memory_buffer_utils` checks out: reserved blocks are never freed or reassigned, so addresses captured by CUDA graphs stay valid, and larger later requests allocate new blocks without touching reserved ones.

Two coordination notes:

- This branch already carries the multistream prev_topk write-back from #16666. If this PR lands first, #16666 should be closed as superseded; if #16666 lands first, the rebase here is trivial.
- #17821 adds a self-sampling GVR dispatch branch inside the decode block this PR removes, and its warmup reads `metadata.enable_heuristic_topk`, which this PR renames. After this merges it needs a rework into a `TopKImplementation` entry rather than a rebase; I can take that on my side.

Remaining points inline.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py (hunk 5 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py:650]: Confirmed. `CUDAGraphRunner`, breakable graphs, and piecewise graphs all run eager forwards before capture. CUDA GVR therefore initializes the C++ `call_once` hardware cache before capture; CuTe GVR does not use this dispatcher.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py (hunk 6 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py:1526]: This is intentional. Dense skip results remain valid priors if a later step enters GVR, including with a lowered admission floor, and updating them avoids stale state.

File: tensorrt_llm/_torch/modules/top_k.py (hunk 7 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py]: Yes. Both paged-MQA producers allocate logits at `indexer_max_seq_len`, with no narrowing before TopK, so `scores.shape[1]` is stable and matches the warmed key.

File: tensorrt_llm/_torch/modules/top_k.py (hunk 8 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py:230]: Clarified in `04d0f9c9e5`. The BF16/FP16 C++ entry intentionally has no split-work auxiliary-buffer interface and rejects widths requiring split work, so TopK should not allocate an unused workspace or duplicate the C++ threshold check.

**Review by longcheng-nv — APPROVED** [review]
